tests/gpflow/posteriors/test_posteriors.py

```diff
@@ -0,0 +1,414 @@
+#  Copyright 2021 The GPflow Contributors. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import warnings
+from inspect import isabstract
+
+import numpy as np
+import pytest
+import tensorflow as tf
+
+import gpflow
+import gpflow.ci_utils
+from gpflow.models.util import inducingpoint_wrapper
+from gpflow.posterior import (
+    AbstractPosterior,
+    FallbackIndependentLatentPosterior,
+    FullyCorrelatedPosterior,
+    IndependentPosteriorMultiOutput,
+    IndependentPosteriorSingleOutput,
+    LinearCoregionalizationPosterior,
+    create_posterior,
+)
+
+INPUT_DIMS = 2
+NUM_INDUCING_POINTS = 3
+
+
+@pytest.fixture(name="q_sqrt_factory", params=[0, 1, 2])
+def _q_sqrt_factory_fixture(request):
+    """
+    When upgrading to Python 3.10, this can be replaced with a match-case statement.
+    """
+    if request.param == 0:
+
+        def fn_0(_, __):
+            return None
+
+        return fn_0
+    elif request.param == 1:
+
+        def fn_1(n_inducing_points, num_latent_gps):
+            # qsqrt: [M, L]
+            return tf.ones((n_inducing_points, num_latent_gps), dtype=tf.float64)
+
+        return fn_1
+    elif request.param == 2:
+
+        def fn_2(n_inducing_points, num_latent_gps):
+            # qsqrt: [L, M, M]
+            return tf.eye(n_inducing_points, batch_shape=[num_latent_gps], dtype=tf.float64)
+
+        return fn_2
+    else:
+        raise NotImplementedError()
+
+
+@pytest.fixture(name="whiten", params=[False, True])
+def _whiten_fixture(request):
+    return request.param
+
+
+@pytest.fixture(name="num_latent_gps", params=[1, 2])
+def _num_latent_gps_fixture(request):
+    return request.param
+
+
+@pytest.fixture(name="output_dims", params=[1, 5])
+def _output_dims_fixture(request):
+    return request.param
+
+
+TESTED_POSTERIORS = set()
+
+
+@pytest.fixture(scope="module", autouse=True)
+def _ensure_all_posteriors_are_tested_fixture():
+    """
+    This fixture ensures that all concrete posteriors have unit tests which compare the predictions
+    from the fused and precomputed code paths. When adding a new concrete posterior class to
+    GPFlow, ensure that it is also tested in this manner.
+    """
+    yield
```

Comment:
I can _imagine_ how this works but I'm not entirely sure and wouldn't know where exactly to start looking to find out - could you add a comment briefly explaining what this does / link to the relevant part of pytest documentation?

Also, I'm assuming you manually tested that this does correctly trigger if you remove one of the tests below, could you confirm that? thanks:)

Reply:
I have added a comment explaining how this is supposed to work, and I have manually tested that it triggers when removing one the tests.